### PR TITLE
Add support for gutenberg image blocks

### LINKED ISSUE #36: An image cannot be selected until Change Image is clicked
Steps to reproduce:

1. Insert an image block and open the media manager
2. Click on any image
3. Once the left panel with the different images is displayed, click on the `Select` button.
4. Nothing happens even if this button is clicked several times.
5. Click on `Change image`.
6. Select the same image again. Now it can be inserted.

When an image is clicked, there's an output error in the JS console. Here's a little GIF that shows it (it's big for GitHub so I'm uploading it to Drive: https://drive.google.com/file/d/1LWQIv5uvyzuaxFRWGI0SsKALCx35HeP8/view)

It seems that the error starts on `cropper.js:183` but triggered in `media-views`:

```js
 sidebar.set( 'details', new Media.view.Attachment.Details( {
      controller: this.controller,
      model: single,
      priority: 80
    } ) );
```
`this.controller` is `undefined` when the image is clicked for the first time yet I'm not 100% sure if that's what is causing the problem when the image is selected but I hope it helps to discover it.

### LINKED ISSUE #37: Image cropper not applying size selection to image blocks
It seems that the image cropper is not working anymore in 3.0 but it could be related to #36 

**How to reproduce**

1. Add an image block and click on _Media Library_
2. Select an image but then click on `Change Image`. Due to #36, you need to do that in order to reproduce this issue.
3. Select now **a different image**.
4. Click on a size on the left side that allows to crop the image and change the cropping area.
5. Select the image and see the results. The selected size is not applied to the block.

**Note**: While writing this I realized that the _Image Size_ selector in the block sidebar does not change when an image is selected. If you try to switch it, the size will change. Maybe the problem is that the image size attribute is not updated when the image is selected. Here's a GIF with the whole process.

Kapture 2020-02-18 at 11 05 50

### PR DESCRIPTION
**Pull request by roborourke**
This also fixes the select button on the modal for WP 5.3.2

Fixes #36
Fixes #37

### CODE REVIEW

File: inc/cropper/src/views/image-edit.js [hunk 1]
```diff
@@ -15,12 +15,27 @@ const ImageEditView = Media.View.extend( {
...
 		// Initial render.
 		this.onUpdate();
 	},
+	setSizeForBlock() {
+		if ( ! wp || ! wp.data ) {
+			return;
+		}
+
+		const selectedBlock = wp.data.select( 'core/block-editor' ).getSelectedBlock();
+		if ( ! selectedBlock || selectedBlock.name !== 'core/image' ) {
```
Comment (roborourke) [inline inc/cropper/src/views/image-edit.js]: probably need to make this work in the other direction too, like `mapBlockAttributesToImageEditSize` or something ...

File: README.md [hunk 2]
```diff
@@ -45,6 +45,29 @@ To disable the feature add the following:
 add_filter( 'hm.smart-media.cropper', '__return_false' );
 ```
 
+The image cropping IU provides support for updating Gutenberg block attributes based on the current selection using the following function:
```
Comment (rmccue) [inline README.md]:
```suggestion
The image cropping UI provides support for updating Gutenberg block attributes based on the current selection using the following function:
```

File: README.md [hunk 3]
```diff
@@ -45,6 +45,29 @@ To disable the feature add the following:
...
+    width: image.width,
+    height: image.height,
+  };
+} );
+```
+
+The function takes 2 parameters:
+
+- `block`: The name of the block to map attributes for
+- `callback`: A function that accepts the image `size` name, an `image` object containing url, width, height, crop data and label for the image size, and lastly the full attachment data object.
```
Comment (rmccue) [inline README.md]:
```suggestion
- `callback`: A function that accepts the image `size` name, an `image` object containing `url`, `width`, `height`, crop data and label for the image size, and lastly the full `attachment` data object.
```

Might need a bit more info?

File: README.md [hunk 4]
```diff
@@ -45,6 +45,29 @@ To disable the feature add the following:
...
+
+The callback should return an object or `null`. Passing `null` will prevent updating the currently selected block.
```
Comment (rmccue) [inline README.md]: What's the shape of the returned object?

File: inc/cropper/src/cropper.js [hunk 5]
```diff
@@ -158,6 +183,38 @@ Media.events.on( 'frame:select:init', frame => {
...
               frame.reset();
             }
+
+            // Update current block if we can map the attachment to attributes.
+            if ( wp && wp.data ) {
+              const selectedBlock = wp.data.select( 'core/block-editor' ).getSelectedBlock();
+              if (
+                ! selectedBlock ||
+                ! window.SmartMedia.ImageBlockAttributeMaps ||
+                ! window.SmartMedia.ImageBlockAttributeMaps[ selectedBlock.name ]
```
Comment (rmccue) [inline inc/cropper/src/cropper.js]: 😢  can we single-line this?

File: README.md (hunk 2 above)
Comment (roborourke) [reply README.md]: good lord 🤦‍♂

File: README.md (hunk 4 above)
Comment (roborourke) [reply README.md]: It's variable, depends on the block's attributes. Not sure how best to show this really, it's such a gnarly system to interact with meaningfully. Maybe there's a better approach tbh, like a filter or something, forgot those were a thing in JS now.

File: inc/cropper/src/cropper.js (hunk 5 above)
Comment (roborourke) [reply inc/cropper/src/cropper.js]: sometimes you complain about long lines! But yes, sure

Comment (rmccue) [reply inc/cropper/src/cropper.js]: I'll do both! I'd prefer we split the `if` into logical groups even if it's duplicative to ensure readability and mental parsing. We could also alias `window.SmartMedia.ImageBlockAttributeMaps` into a shorter var.

File: inc/cropper/src/views/image-edit.js [hunk 6]
```diff
@@ -15,12 +16,43 @@ const ImageEditView = Media.View.extend( {
...
+		}
+
+		const sizeForBlock = applyFilters(
+			`smartmedia.cropper.selectSizeFromBlockAttributes.${ selectedBlock.name.replace( /\W+/g, '.' ) }`,
+			null,
+			selectedBlock
+		);
+
+		const size = applyFilters(
+			`smartmedia.cropper.selectSizeFromBlockAttributes`,
```
Comment (joehoyle) [inline inc/cropper/src/views/image-edit.js]: Looks like no need for templated string here.

Comment (roborourke) [reply inc/cropper/src/views/image-edit.js]: ah! cheers, left over copy/pasta

**Pull request merged**